OpenSim/Common/DataAdapter.h

```diff
@@ -0,0 +1,225 @@
+               For file adapters, .*/
+    static DataAdapter* createAdapter(const std::string& identifier);
+
+    /** Register a concrete DataAdapter by its unique string identifier.
+        Registration permits access to the required concrete adapter by
+        identifier lookup. As such, identifiers must be unique, but adapters may
+        be registered with multiple identifiers. For example, a data file may 
+        have multiple valid extensions (e.g. ".jpg: and ".jpeg") in which case
+        both extensions would be valid identifiers for the same adapter. If an
+        identifier is already in use an Exception is thrown. 
+    @param[in] identifier string used to uniquely identify the required
+               concrete adapter to process a source/sink. For file adapters,
+               register unique file extension(s) in order for OpenSim to read/
+               write different file formats.
+    @param[in] adapter the concrete DataAdapter required to process the type
+                       (format) of the data for the source/sink.*/
+    static void registerDataAdapter(const std::string& identifier,
+        const DataAdapter& adapter);
+
+    /** Convenience construction of a DataTable from a data source identified
+        by name and its type.
+    @param[in] source           string identifying the source to be opened
+    @param[in] sourceTypeName   string identifying the source type used to
+                                determine which concrete adapter is used.
+    @return DataTable           the resulting in memory data from the source
+        */
+    AbstractDataTable* createDataTable(const std::string& source,
+        const std::string& sourceTypeName);
+
+    /** Direct DataAdpater data flow into the provided DataTable. */
+//TODO? DataTable& operator>>(DataTable& table);
+
+    /** Direct DataTable data out through the DataAdapter. */
+//TODO? DataAdapter& operator<<(const DataTable& table);
+
+    /** Adapter has access to the data source for reading and/or writing? */
+    bool hasDataAccess() const;
+
+    /** Determine read/write access of the adapter to the data source.*/
+    std::ios_base::openmode getAccessMode() const;
+
+    /** Open the data source to enable access to read and/or writing of 
+        data according to the access mode. If successful returns true.
+        hasDataAccess() will also return true.*/
+    bool openDataSource();
+
+    /** Close the connection used to access the data source. 
+        Returns true if successfully closed access to data. hasDataAccess() 
+        also return false.*/
+    bool closeDataSource();
+
+    /** Give the adapter an opportunity to prepare itself for reading data
+        into a DataTable. Adapter keeps a reference to the DataTable for
+        all subsequent read access.
+    @param[out] table  The DataTable to be filled by the DataAdapter. */
+    void prepareForReading(AbstractDataTable& table);
```

Comment:
How does this know that the AbstractDataTable has the right kind of data elements? Can a DataAdapter fill in any kind of elements, or does use of a particular Adapter imply a type T such that it can handle only `DataTable<T>` objects?